File: 2014-10-CSharp-Fundamentals-1/2. Primitive Data Types and Variables/FloatingPointTypes/FloatingPointTypes.cs
using System;

class FloatingPointTypes
{
    static void Main()
    {
        float floatPI = 3.141592653589793238f;
        double doublePI = 3.141592653589793238;
        Console.WriteLine("Float PI is: {0}", floatPI);
        Console.WriteLine("Double PI is: {0}", doublePI);
        Console.WriteLine();

        // Example of loss of precision (float to double)
        double a = 1.0f;
        double b = 0.33f;
        double expectedSum = 1.33f;
        bool equal = (a + b == expectedSum); // False!!!
        Console.WriteLine(
            "Float calculation: a+b={0}  expected={1}  equal={2}",
            a + b, expectedSum, equal);
        Console.WriteLine();

        // Example of comparison abnormality (double)
        a = 0.1;
        b = 0.2;
        double sum = a + b;
        expectedSum = 0.3;
        equal = (sum == expectedSum);
        Console.WriteLine(
            "Double calculation: a+b={0}  expected={1}  equal={2}",
            sum, expectedSum, equal);
        Console.WriteLine("sum (without rounding) = " +
            System.ComponentModel.TypeDescriptor.GetConverter(sum).ConvertTo(sum, typeof(string)));
        Console.WriteLine();

        // Decimal numbers do not have comparison abnormalities
        decimal aDecimal = 1.0M;
        decimal bDecimal = 0.33M;
        decimal sumDecimal = 1.33M;
        bool equalDecimal = (aDecimal + bDecimal == sumDecimal);
        Console.WriteLine(
			"Decimal calculation: a+b={0}  sum={1}  equal={2}", 
			aDecimal + bDecimal, sumDecimal, equalDecimal);
    }
}

File: 2015-10-Java-Fundamentals-1/02. Data-types-and-variables/demos/src/FloatingPointTypes.java
import java.math.BigDecimal;
import java.text.MessageFormat;

public class FloatingPointTypes {
    public static void main(String[] args) {
        float floatPI = 3.141592653589793238f;
        double doublePI = 3.141592653589793238;
        WriteLine("Float PI is: {0}", floatPI);
        WriteLine("Double PI is: {0}", doublePI);

        // Example of comparison abnormality
        double a = 1.0f;
        double b = 0.33f;
        double sum = 1.33f;
        boolean equal = (a + b == sum);
        WriteLine("Float calculation: a+b={0}  sum={1}  equal={2}",
                a + b, sum, equal);

        // Decimal numbers do not have comparison abnormalities
        BigDecimal aDecimal =   BigDecimal.valueOf(1.0);
        BigDecimal bDecimal =   BigDecimal.valueOf(0.33);
        BigDecimal sumDecimal = BigDecimal.valueOf(1.33);
        boolean equalDecimal = (aDecimal.add(bDecimal).equals(sumDecimal));
        WriteLine("Decimal calculation: a+b={0}  sum={1}  equal={2}",
                aDecimal.add(bDecimal), sumDecimal, equalDecimal);

    }

    public static void WriteLine(String format, Object... obj) {
        MessageFormat form = new MessageFormat(format);
        System.out.println(form.format(obj));
    }
}
